### PR TITLE
build middleware as checker for authorization

### PR DESCRIPTION
**Pull request by ife-jeremiah**

### CODE REVIEW

File: server/src/modules/property/property.route.ts [hunk 1]
```diff
@@ -9,22 +9,23 @@ import {
...
+router.get('/me', authorize(), findPropertyByUserId);
 
 router
   .route('/:id')
-  .get(findPropertiesById)
-  .patch(processRequestBody(updatePropertySchema.body), updatePropertyById)
-  .delete(deletePropertyById);
+  .get(authorize(), findPropertiesById)
+  .patch(authorize(), processRequestBody(updatePropertySchema.body), updatePropertyById)
+  .delete(authorize('ADMIN'), deletePropertyById);
```
Comment (adamilare) [inline server/src/modules/property/property.route.ts:29]: I think we are going to be passing the 'roleID' not the 'roleName'

File: server/src/middleware/authorize.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import { apiError } from '../../src/utils/response';
+import { findById } from '../../src/modules/profile/profile.service';
+
+export function authorize(roles: string | string[] | null = null) {
+  if (typeof roles === 'string') {
+    roles = [roles];
+  }
+
+  return [
+    jwt({ secret: JWT_SECRET, algorithms: ['HS256'] }),
```
[the remark this answers is not in the document]
Comment (ife-jeremiah) [reply server/src/middleware/authorize.ts:13]: it's not so different from what `deserializeUser` is doing, but it's an authentication middleware on the controller level.

File: server/src/middleware/authorize.ts
[the remark this answers is not in the document]
Comment (ife-jeremiah) [reply server/src/middleware/authorize.ts:1]: i don't get for this part sir.

File: server/src/modules/property/property.route.ts (hunk 1 above)
Comment (ife-jeremiah) [reply server/src/modules/property/property.route.ts:29]: oh, okay, we could sync with that.

File: server/src/middleware/authorize.ts
[the remark this answers is not in the document]
Comment (adamilare) [reply server/src/middleware/authorize.ts:1]: We need to check beyond if user exists. The edit checks if any of their present role-types qualify to access the endpoint and complete the transaction.

When argument is null, it means being a user is enough for the transaction

[the remark this answers is not in the document]
Comment (ife-jeremiah) [reply server/src/middleware/authorize.ts:1]: hmm, very true. thank you.

**Pull request merged**